internal/gcsx/file_cache_reader_test.go

```diff
@@ -81,3 +98,209 @@ func (t *FileCacheReaderTest) TestNewFileCacheReader() {
+	end1 := util.MiB
+	assert.Less(t.T(), end1, int(t.object.Size))
+	// First call from offset 0 - sequential read
+	buf := make([]byte, end1-start1)
+	_, cacheHit, err := t.reader.tryReadingFromFileCache(t.ctx, buf, int64(start1))
+	assert.NoError(t.T(), err)
+	assert.False(t.T(), cacheHit)
+	assert.Equal(t.T(), buf, testContent[start1:end1])
+	start2 := 3*util.MiB + 4
+	end2 := start2 + util.MiB
+	buf2 := make([]byte, end2-start2)
+
+	// Assuming start2 offset download in progress
+	_, cacheHit, err = t.reader.tryReadingFromFileCache(t.ctx, buf2, int64(start2))
+
+	assert.NoError(t.T(), err)
+	assert.True(t.T(), cacheHit)
+	assert.Equal(t.T(), buf2, testContent[start2:end2])
+	t.mockBucket.AssertExpectations(t.T())
+}
+
+func (t *fileCacheReaderTest) Test_ReadAt_SequentialRangeRead() {
+	testContent := testutil.GenerateRandomBytes(int(t.object.Size))
+	rd := &fake.FakeReader{ReadCloser: getReadCloser(testContent)}
+	t.mockNewReaderWithHandleCallForTestBucket(0, t.object.Size, rd)
+	t.mockBucket.On("Name").Return("test-bucket")
+	start := 0
+	end := 10
+	assert.Less(t.T(), end, int(t.object.Size))
+	buf := make([]byte, end-start)
+
+	readerResponse, err := t.reader.ReadAt(t.ctx, buf, int64(start))
+
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), readerResponse.DataBuf, testContent[start:end])
+	t.mockBucket.AssertExpectations(t.T())
+}
+
+func (t *fileCacheReaderTest) Test_ReadAt_RandomReadNotStartWithZeroOffsetWhenCacheForRangeReadIsFalse() {
+	t.reader.cacheFileForRangeRead = false
+	start := 5
+	end := 10
+	t.mockBucket.On("Name").Return("test-bucket")
+	buf := make([]byte, end-start)
+	readerResponse, err := t.reader.ReadAt(t.ctx, buf, int64(start))
+	assert.True(t.T(), errors.Is(err, FallbackToAnotherReader))
+	assert.Zero(t.T(), readerResponse.Size)
+	job := t.jobManager.CreateJobIfNotExists(t.object, t.mockBucket)
+	jobStatus := job.GetStatus()
+	assert.True(t.T(), jobStatus.Name == downloader.NotStarted)
+
+	readerResponse, err = t.reader.ReadAt(t.ctx, buf, int64(start))
+
+	assert.True(t.T(), errors.Is(err, FallbackToAnotherReader))
+	assert.Zero(t.T(), readerResponse.Size)
+	t.mockBucket.AssertExpectations(t.T())
+}
+
+func (t *fileCacheReaderTest) Test_ReadAt_RandomReadNotStartWithZeroOffsetWhenCacheForRangeReadIsTrue() {
+	t.reader.cacheFileForRangeRead = true
+	testContent := testutil.GenerateRandomBytes(int(t.object.Size))
+	start := 5
+	end := 10
+	rd := &fake.FakeReader{ReadCloser: getReadCloser(testContent)}
+	// Mock for download job's NewReader call
+	t.mockNewReaderWithHandleCallForTestBucket(0, t.object.Size, rd)
+	t.mockBucket.On("Name").Return("test-bucket")
+	buf := make([]byte, end-start)
+
+	readerResponse, err := t.reader.ReadAt(t.ctx, buf, int64(start))
+
+	assert.True(t.T(), errors.Is(err, FallbackToAnotherReader))
+	assert.Zero(t.T(), readerResponse.Size)
+	job := t.jobManager.GetJob(t.object.Name, t.mockBucket.Name())
+	assert.True(t.T(), job == nil || job.GetStatus().Name == downloader.Downloading)
```

Comment:
Can we validate the non-nil cacheHandle also just to make sure cache got triggered?

As job == nil can happen when cache-path is not triggered.